engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterSceneLifeCycle.mm

```diff
@@ -0,0 +1,390 @@
+    if (result) {
+      consumedByPlugin = YES;
+    }
+  }
+  if (!consumedByPlugin) {
+    BOOL result =
+        [[self applicationLifeCycleDelegate] sceneFallbackContinueUserActivity:userActivity];
+    if (result) {
+      consumedByPlugin = YES;
+    }
+  }
+  return consumedByPlugin;
+}
+
+#pragma mark - Performing tasks
+
+- (BOOL)windowScene:(UIWindowScene*)windowScene
+    performActionForShortcutItem:(UIApplicationShortcutItem*)shortcutItem
+               completionHandler:(void (^)(BOOL succeeded))completionHandler {
+  [self updateEnginesInScene:windowScene];
+
+  BOOL consumedByPlugin = NO;
+  for (FlutterEngine* engine in _engines.allObjects) {
+    BOOL result = [engine.sceneLifeCycleDelegate windowScene:windowScene
+                                performActionForShortcutItem:shortcutItem
+                                           completionHandler:completionHandler];
+    if (result) {
+      consumedByPlugin = YES;
+    }
+  }
+  if (!consumedByPlugin) {
+    BOOL result = [[self applicationLifeCycleDelegate]
+        sceneFallbackPerformActionForShortcutItem:shortcutItem
+                                completionHandler:completionHandler];
+    if (result) {
+      consumedByPlugin = YES;
+    }
+  }
+  return consumedByPlugin;
+}
+@end
+
+@implementation FlutterEnginePluginSceneLifeCycleDelegate {
+  // Weak references to registered plugins.
+  NSPointerArray* _delegates;
+}
+
+- (instancetype)init {
+  if (self = [super init]) {
+    _delegates = [NSPointerArray weakObjectsPointerArray];
+  }
+  return self;
+}
+
+- (void)addDelegate:(NSObject<FlutterSceneLifeCycleDelegate>*)delegate {
+  [_delegates addPointer:(__bridge void*)delegate];
+
+  // NSPointerArray is clever and assumes that unless a mutation operation has occurred on it that
+  // has set one of its values to nil, nothing could have changed and it can skip compaction.
+  // That's reasonable behaviour on a regular NSPointerArray but not for a weakObjectPointerArray.
+  // As a workaround, we mutate it first. See: http://www.openradar.me/15396578
+  [_delegates addPointer:nil];
+  [_delegates compact];
+}
+
+#pragma mark - Connecting and disconnecting the scene
+
+- (BOOL)scene:(UIScene*)scene
+    willConnectToSession:(UISceneSession*)session
+                 options:(UISceneConnectionOptions*)connectionOptions {
+  BOOL consumedByPlugin = NO;
+  for (NSObject<FlutterSceneLifeCycleDelegate>* delegate in _delegates.allObjects) {
+    if ([delegate respondsToSelector:_cmd]) {
+      // If this event has already been consumed by a plugin, send the event with nil options.
+      if (consumedByPlugin) {
+        [delegate scene:scene willConnectToSession:session options:nil];
+        continue;
+      } else if ([delegate scene:scene willConnectToSession:session options:connectionOptions]) {
+        // Only allow one plugin to process this event.
+        consumedByPlugin = YES;
+      }
+    }
```

Comment:
Optional: it might be easier to understand this as just:

```objc
// Only allow one plugin to process the connection options.
if ([delegate scene:scene willConnectToSession:session options:(consumedByPlugin ? nil : connectionOptions)]) {
  consumedByPlugin = YES;
}
```